File: src/Cell.kt
import Printer.Companion.printError

class Cell(val x: Int, val y: Int, var state: State){

    private lateinit var nextState: State

    val isOnTop = y == 0
    val isOnBottom = y == BOARD_SIZE - 1
    val isOnStart = x == 0
    val isOnEnd = x == BOARD_SIZE - 1

    fun prepareSimulate(aroundCells: List<Cell>) {
        if (aroundCells.size > 8) {
            printError()
            return
        }

        nextState = when (state) {
            State.Living -> when (aroundCells.filter{ it.state == State.Living }.size) {
                2, 3 -> State.Living
                else -> State.Dead
            }
            State.Dead -> when (aroundCells.filter{ it.state == State.Living }.size) {
                3 -> State.Living
                else -> State.Dead
            }
        }
    }

    fun simulate() {
        state = nextState
    }

    enum class State {
        Living {
            override fun getMark() = "■ "
        },
        Dead {
            override fun getMark() = "□ "
        };
        abstract fun getMark(): String
    }
}

File: Nim/cell.nim
import sequtils

type CellState* = enum
    DEAD
    ALIVE

type Cell* = ref object
    x: int
    y: int
    nextState: CellState
    currentState: CellState

method getX*(cell: Cell): int {.base.} = cell.x
method getY*(cell: Cell): int {.base.} = cell.y

method setCurrentState*(cell: Cell, state: CellState) {.base.} =
    cell.currentState = state
    

method getStateMark*(cell: Cell): string {.base.} = 
    result = case cell.currentState:
        of ALIVE:
            "■ "
        else:
            "□ "

method toNextGen*(cell: Cell) {.base.} =
    cell.currentState = cell.nextState

method init*(cell: Cell, x: int, y: int, currentState: CellState) {.base.} = 
    cell.x = x
    cell.y = y
    cell.currentState = currentState

method prepareCell*(cell: Cell, aroundCells: seq[Cell]) {.base.} =
    let aliveCount = aroundCells.filterIt(it.currentState == ALIVE).len()
    if cell.currentState == ALIVE:
        cell.nextState = case aliveCount:
            of 2, 3:
                ALIVE
            else:
                DEAD
    else:
        cell.nextState = case aliveCount:
            of 3:
                ALIVE
            else:
                DEAD

method isOnTop*(cell: Cell) : bool {.base.} = cell.y == 0
method isOnBottom*(cell: Cell, boardSize: int) : bool {.base.} = cell.y == boardSize - 1
method isOnStart*(cell: Cell) : bool {.base.} = cell.x == 0
method isOnEnd*(cell: Cell, boardSize: int): bool {.base.} = cell.x == boardSize - 1
